### DISCUSSION TITLE
Is it possible to bruteforce OTP on SSH?

### DISCUSSION THREAD

**Discussion by fungiboletus** [warp-tech/warpgate#1079] [Q&A]
Is there a mechanism to prevent brute-forcing OTP when login through SSH? And shouldn't the One-time password be requested after a successful password?

**Comment by Kiprey** [comment]
I had the same question and wrote an exploit to brute force the SSH OTP.

What I found is that all OTP validation seems to be handled centrally by Warpgate. Even when I ran 50 concurrent SSH attempts, the brute force rate I could reach locally was still below 200 OTPs per second.

This is probably because Warpgate processes OTPs on a single core, and each SSH attempt takes around 100 ms to test one OTP. At that speed, successfully brute forcing the OTP within a 30 second x 2 window would be extremely difficult.

**Reply by fungiboletus** [reply]
Couldn't you keep bruteforcing forever? Quick maths show that a few days should be enough at this rate.

**Reply by Eugeny** [reply]
@fungiboletus you need to either brute force the 6-digit code within 30 seconds, or brute force the 256-bit secret key

**Reply by fungiboletus** [reply]
@Eugeny can't I just send many random 6 digits code forever and hope for the best? One will eventually be valid. Or does it has an exponential rate limit mechanism?
